src/bulk-load.js

```diff
@@ -260,4 +275,87 @@ module.exports = class BulkLoad extends EventEmitter {
     }
     return tBuf.data;
   }
-};
+
+  createDoneToken() {
+    // It might be nice to make DoneToken a class if anything needs to create them, but for now, just do it here
+    const tBuf = new WritableTrackingBuffer(this.options.tdsVersion < '7_2' ? 9 : 13);
+    tBuf.writeUInt8(TOKEN_TYPE.DONE);
+    const status = DONE_STATUS.FINAL;
+    tBuf.writeUInt16LE(status);
+    tBuf.writeUInt16LE(0); // CurCmd (TDS ignores this)
+    tBuf.writeUInt32LE(0); // row count - doesn't really matter
+    if (this.options.tdsVersion >= '7_2') {
+      tBuf.writeUInt32LE(0); // row count is 64 bits in >= TDS 7.2
+    }
+    return tBuf.data;
+  }
+
+  // This method switches the BulkLoad object into streaming mode and returns
+  // a stream.Writable for streaming rows to the server.
+  getRowStream() {
+    if (this.firstRowWritten) {
+      throw new Error('BulkLoad cannot be switched to streaming mode after first row has been written using addRow().');
+    }
+    if (this.executionStarted) {
+      throw new Error('BulkLoad cannot be switched to streaming mode after execution has started.');
+    }
+    this.streamingMode = true;
+    return this.rowToPacketTransform;
+  }
+
+  getMessageStream() {
+    const message = new Message({ type: PACKET_TYPE.BULK_LOAD });
+    this.rowToPacketTransform.pipe(message);
+    return message;
+  }
+}
+
+module.exports = BulkLoad;
+
+// A transform that converts rows to packets.
+class RowTransform extends Transform {
+  columnMetadataWritten: boolean;
+  bulkLoad: BulkLoad;
+  mainOptions: $PropertyType<BulkLoad, 'options'>;
+  columns: $PropertyType<BulkLoad, 'columns'>;
+
+  constructor(bulkLoad: BulkLoad) {
+    super({ writableObjectMode: true });
+
+    this.bulkLoad = bulkLoad;
+    this.mainOptions = bulkLoad.options;
+    this.columns = bulkLoad.columns;
+
+    this.columnMetadataWritten = false;
+  }
+
+  _transform(row, encoding, callback) {
+    if (!this.columnMetadataWritten) {
+      this.push(this.bulkLoad.getColMetaData());
+      this.columnMetadataWritten = true;
+    }
+
+    const buf = new WritableTrackingBuffer(10, 'ucs2', true);
```

Comment:
10 bytes are a bit small for the start. Even when writing only 2 integer values, we already need 11 bytes and `WritableTrackingBuffer.makeRoomFor()` will copy and expand the buffer to 128 bytes.

Reply:
I changed this to 64 byte. How do you feel about that size? 🙇